server/src/main/java/org/opensearch/transport/InboundPipeline.java

```diff
@@ -124,95 +121,12 @@ public void doHandleBytes(TcpChannel channel, ReleasableBytesReference reference
-                        }
-                    }
-                    fragments.clear();
-                }
-            }
-        }
-    }
-
-    private void forwardFragments(TcpChannel channel, ArrayList<Object> fragments) throws IOException {
-        for (Object fragment : fragments) {
-            if (fragment instanceof Header) {
-                assert aggregator.isAggregating() == false;
-                aggregator.headerReceived((Header) fragment);
-            } else if (fragment == InboundDecoder.PING) {
-                assert aggregator.isAggregating() == false;
-                messageHandler.accept(channel, PING_MESSAGE);
-            } else if (fragment == InboundDecoder.END_CONTENT) {
-                assert aggregator.isAggregating();
-                try (InboundMessage aggregated = aggregator.finishAggregation()) {
-                    statsTracker.markMessageReceived();
-                    messageHandler.accept(channel, aggregated);
-                }
-            } else {
-                assert aggregator.isAggregating();
-                assert fragment instanceof ReleasableBytesReference;
-                aggregator.aggregate((ReleasableBytesReference) fragment);
-            }
-        }
-    }
-
-    private boolean endOfMessage(Object fragment) {
-        return fragment == InboundDecoder.PING || fragment == InboundDecoder.END_CONTENT || fragment instanceof Exception;
-    }
-
-    private ReleasableBytesReference getPendingBytes() {
-        if (pending.size() == 1) {
-            return pending.peekFirst().retain();
-        } else {
-            final ReleasableBytesReference[] bytesReferences = new ReleasableBytesReference[pending.size()];
-            int index = 0;
-            for (ReleasableBytesReference pendingReference : pending) {
-                bytesReferences[index] = pendingReference.retain();
-                ++index;
-            }
-            final Releasable releasable = () -> Releasables.closeWhileHandlingException(bytesReferences);
-            return new ReleasableBytesReference(CompositeBytesReference.of(bytesReferences), releasable);
-        }
-    }
-
-    private void releasePendingBytes(int bytesConsumed) {
-        int bytesToRelease = bytesConsumed;
-        while (bytesToRelease != 0) {
-            try (ReleasableBytesReference reference = pending.pollFirst()) {
-                assert reference != null;
-                if (bytesToRelease < reference.length()) {
-                    pending.addFirst(reference.retainedSlice(bytesToRelease, reference.length() - bytesToRelease));
-                    bytesToRelease -= bytesToRelease;
-                } else {
-                    bytesToRelease -= reference.length();
-                }
+        for (InboundBytesHandler handler : protocolBytesHandlers) {
+            if (handler.canHandleBytes(reference)) {
+                handler.doHandleBytes(channel, reference, isClosed, pending, decoder, aggregator, statsTracker);
```

Comment:
Do you mean for each reference, `doHandleBytes` would be called multiple times?

Reply:
I think we would want to detect protocol for each `doHandleBytes` call since we would be implementing protobuf (a new protocol) partially for one API for one example. Would that not require us to detect the protocol for each bytes reference?

Reply:
> Do you mean for each reference, doHandleBytes would be called multiple times?

Yes, for the same message (so I suspect we should not mixed protocols here)

> I think we would want to detect protocol for each doHandleBytes call since we would be implementing protobuf (a new protocol) partially for one API for one example.

I believe the message may not be received all at once, but in fragments, and `doHandleBytes` will be called multiple times with new fragments  (this is why we have decoder to decode fragments and aggregator to stich them together)

Reply:
Posting the result of discussion here: the protocol should be determined only on first byte reference for a fragmented message. The new commit in this PR addresses that by having a currentHandler and checking if its non-null, then use the same handler.

Reply:
Let me know if this looks good.